image-builder/build.sh

```diff
@@ -0,0 +1,187 @@
+  CALCULATED_SHA256="$(sha256sum "$TEMPDIR/CentOS-7-x86_64-GenericCloud-$CENTOS7_BUILD.qcow2.xz" | cut -f1 -d ' ')"
+  if [[ "$CALCULATED_SHA256" != "$EXPECTED_SHA256" ]]; then
+    echo " * SHA256 digest verification failed. '$CALCULATED_SHA256' != '$EXPECTED_SHA256'"
+    exit 1
+  fi
+
+  echo " * Decompressing"
+  unxz --keep "$TEMPDIR/CentOS-7-x86_64-GenericCloud-$CENTOS7_BUILD.qcow2.xz"
+  mv "$TEMPDIR/CentOS-7-x86_64-GenericCloud-$CENTOS7_BUILD.qcow2" "$SCRIPT_DIR/downloads/CentOS-7-x86_64-GenericCloud.qcow2"
+}
+
+get_debian9_image() {
+  DEBIAN_CD_SIGNING_KEY_FINGERPRINT="DF9B9C49EAA9298432589D76DA87E80D6294BE9B"
+
+  echo " * Downloading vanilla Debian image."
+  wget "https://cdimage.debian.org/cdimage/openstack/current-9/debian-9-openstack-amd64.qcow2" -P "$TEMPDIR"
+
+  echo " * Verifying GPG signature"
+  wget --quiet "https://cdimage.debian.org/cdimage/openstack/current-9/SHA512SUMS" -O "$TEMPDIR/Debian-SHA512SUMS"
+  wget --quiet "https://cdimage.debian.org/cdimage/openstack/current-9/SHA512SUMS.sign" -O "$TEMPDIR/Debian-SHA512SUMS.sign"
+  gpg2 --quiet --recv-keys "$DEBIAN_CD_SIGNING_KEY_FINGERPRINT"
+  gpg2 --quiet --verify "$TEMPDIR/Debian-SHA512SUMS.sign"
+
+  echo " * Verifying SHA512 digest"
+  EXPECTED_SHA512="$(grep 'debian-9-openstack-amd64.qcow2$' < "$TEMPDIR/Debian-SHA512SUMS" | cut -f1 -d ' ')"
+  CALCULATED_SHA512="$(sha512sum "$TEMPDIR/debian-9-openstack-amd64.qcow2" | cut -f1 -d ' ')"
+  if [[ "$CALCULATED_SHA512" != "$EXPECTED_SHA512" ]]; then
+    echo " * SHA512 digest verification failed. '$CALCULATED_SHA512' != '$EXPECTED_SHA512'"
+    exit 1
+  fi
+
+  echo " * Finalizing"
+  mv "$TEMPDIR/debian-9-openstack-amd64.qcow2" "$SCRIPT_DIR/downloads/debian-9-openstack-amd64.qcow2"
+}
+
+case $TARGET_OS in
+  coreos)
+    CLEAN_IMAGE="$SCRIPT_DIR/downloads/coreos_production_vmware_image.original.vmdk"
+    ROOTFS_PARTITION="/dev/sda9"
+    if [[ ! -f "$CLEAN_IMAGE" ]]; then
+      get_coreos_image
+    fi
+    ;;
+  centos7)
+  CLEAN_IMAGE="$SCRIPT_DIR/downloads/CentOS-7-x86_64-GenericCloud.qcow2"
+    ROOTFS_PARTITION="/dev/sda1"
+    if [[ ! -f "$CLEAN_IMAGE" ]]; then
+      get_centos7_image
+    fi
+    ;;
+  debian9)
+    CLEAN_IMAGE="$SCRIPT_DIR/downloads/debian-9-openstack-amd64.qcow2"
+    ROOTFS_PARTITION="/dev/sda1"
+    if [[ ! -f "$CLEAN_IMAGE" ]]; then
+      get_debian9_image
+    fi
+    ;;
+  *)
+    usage
+    exit 1
+esac
+
+echo " * Verifying/Downloading kubernetes"
+./download_kubernetes.sh --release "$K8S_RELEASE"
+
+echo " * Mouting the image"
+cp "$CLEAN_IMAGE" "$TEMPDIR/work-in-progress-image"
+sudo guestmount -a "$TEMPDIR/work-in-progress-image" -m "$ROOTFS_PARTITION" "$TARGETFS"
+
+echo " * Copying kubernetes binaries"
+sudo mkdir -p "$TARGETFS/opt/bin/"
+sudo cp "$SCRIPT_DIR/downloads/kubeadm-$K8S_RELEASE" "$TARGETFS/opt/bin/kubeadm"
+sudo cp "$SCRIPT_DIR/downloads/kubectl-$K8S_RELEASE" "$TARGETFS/opt/bin/kubectl"
```

Comment:
For CoreOS we use the hyperkube image instead of kubelet/kubeadm:

https://github.com/kubermatic/machine-controller/blob/master/pkg/userdata/coreos/userdata.go#L170-L222